### PR TITLE
Feature: Playing around with a (generic) paged word-press reader

### PR DESCRIPTION
**Pull request by pavax**
This Branch integrates localhero (based on a word-press api) and indicates on how we could integrate other word-press based api's with ease (famous last words)

### CODE REVIEW

File: integrations/localhero/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+RUN mkdir -p /workspace
+WORKDIR /workspace
+COPY pom.xml /workspace
+RUN mvn dependency:go-offline
+
+COPY src /workspace/src
+RUN mvn package -DskipTests
+
+# Step : Package image
+FROM openjdk:11-jdk-slim
```
Comment (pavax) [inline integrations/localhero/Dockerfile:12]: @lugipfupf sorry that I downgraded to version 11 - I wanted to align it with the other service (sync-service) that I already had -> if you have the time, we can of course upgrade both services to the same version

File: integrations/localhero/src/main/java/ch/stayhome/integrations/localhero/job/LocalHeroBatchConfig.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+				.incrementer(new RunIdIncrementer())
+				.start(this.parallelFlow())
+				.build();
+	}
+
+	private Step parallelFlow() {
+		final List<SourceConfig> sourceConfigs = this.localHeroProperties.getSources();
+		return this.stepBuilderFactory.get("importing-steps")
+				.flow(new FlowBuilder<Flow>("importing-flows")
+						.split(this.parallelTaskExecutor())
```
Comment (pavax) [inline integrations/localhero/src/main/java/ch/stayhome/integrations/localhero/job/LocalHeroBatchConfig.java:83]: @lugipfupf FYI: this is the cool part -> we execute the provided flows in parallel (depending on the configured amount of parallel worker threads 'setConcurrencyLimit' in the executor

File: integrations/localhero/src/main/java/ch/stayhome/integrations/localhero/job/LocalHeroBatchConfig.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+		return new FlowBuilder<Flow>("flow-" + sourceConfig.getKey())
+				.start(step(sourceConfig))
+				.build();
+	}
+
+	private Step step(SourceConfig sourceConfig) {
+		final LocalHeroChApi localHeroChApi = this.buildApi(sourceConfig);
+		return stepBuilderFactory.get("import-step-" + sourceConfig.getKey())
+				.<LocalHeroPost, StayHomeEntry>chunk(localHeroProperties.getChunkSize())
+				.faultTolerant()
```
Comment (pavax) [inline integrations/localhero/src/main/java/ch/stayhome/integrations/localhero/job/LocalHeroBatchConfig.java:101]: @lugipfupf we don't want the job to fail (exit) if a single entry failed -> let's skip up to ```X entries``` of the ```given Exceptions``` till the batch-job is considered as failed

File: integrations/localhero/README.md [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+## Development
+Instead of spamming local-hero.ch with requests while developing, there is a local server located in the `mock-server`
+folder that provides a handful posts copied from the actual site.
+
+```bash
+npm install
+npm run json-server
+```
+
+Open `http://localhost:3001` in any browser to find the routes (currently there is only one...). The `local` SpringBoot
```
Comment (pavax) [inline integrations/localhero/README.md]: @lugipfupf  this is where I could need your help to get this mock server ap and running again with the new paginated approach

File: integrations/localhero/src/main/java/ch/stayhome/integrations/localhero/job/LocalHeroPostProcessor.java [hunk 5]
```diff
@@ -9,30 +16,48 @@
...
+				.categories(determineCategories(item))
+				.delivery(this.sourceConfig.getPlace())
+				.contact("")
+				.email("")
+				.phone("")
+				.ttl(defaultTTL)
+				.build();
+	}
+
+	private String determineCategories(LocalHeroPost item) {
```
Comment (pavax) [inline integrations/localhero/src/main/java/ch/stayhome/integrations/localhero/job/LocalHeroPostProcessor.java:52]: @lugipfupf This is where I tried to resolve the category ids from localhero to get the name and then join everything into the categories field that the sync-service needs

File: integrations/localhero/src/main/java/ch/stayhome/integrations/localhero/config/LocalHeroProperties.java [hunk 6]
```diff
@@ -9,31 +10,45 @@
...
 @Data
 public class LocalHeroProperties {
-    @NotBlank(message = "No cron expression has been defined in application.yml")
-    private String scrapeCron;
 
-    @NotEmpty
-    private List<String> sourceUrls;
-    
-    @NotBlank(message = "No rest route has been defined in application.yml")
-    private String restRoute;
```
Comment (pavax) [inline integrations/localhero/src/main/java/ch/stayhome/integrations/localhero/config/LocalHeroProperties.java]: @lugipfupf  I was not able to use this in combination with the index.php?restRoute=XXXX and the pagination -> Thus I hard-coded the value directly into the LocalHeroChApi.java. It's also where I defined the route to resolve the categories ```/wp-json/wp/v2/categories/```

File: integrations/localhero/src/main/java/ch/stayhome/integrations/localhero/infrastructure/feign/LocalHeroChApi.java [hunk 7]
```diff
@@ -1,12 +1,18 @@
...
 import feign.Param;
 import feign.RequestLine;
 
-import java.util.List;
 
 public interface LocalHeroChApi {
-    @RequestLine("GET /index.php?rest_route={restRoute}")
-    List<LocalHeroPost> findAll(@Param("restRoute") String restRoute);
+
+	@RequestLine("GET /wp-json/wp/v2/posts?rest_route={restRoute}&page={page}&per_page={size}")
```
Comment (pavax) [inline integrations/localhero/src/main/java/ch/stayhome/integrations/localhero/infrastructure/feign/LocalHeroChApi.java]: @lugipfupf I was not able to use this in combination with the index.php?restRoute=XXXX and the pagination -> Thus I hard-coded the value directly into the LocalHeroChApi.java. It's also where I defined the route to resolve the categories /wp-json/wp/v2/categories/

File: integrations/localhero/Dockerfile (hunk 1 above)
Comment (lugipfupf) [reply integrations/localhero/Dockerfile:12]: Not that it really matters to me. As long as it's not a JEE from aeons ago ;-)

File: integrations/localhero/pom.xml [hunk 8]
```diff
@@ -47,7 +49,10 @@
             <artifactId>feign-slf4j</artifactId>
             <version>${feign.version}</version>
         </dependency>
-
+        <dependency>
+            <groupId>com.h2database</groupId>
```
Comment (lugipfupf) [inline integrations/localhero/pom.xml]: Why do we need a database? For the batch? Afaik, it creates a map based storage when no datasource is defined. I went without it on purpose.

File: integrations/localhero/src/main/java/ch/stayhome/integrations/localhero/infrastructure/feign/LocalHeroChApi.java [hunk 9]
```diff
@@ -1,12 +1,18 @@
...
 import feign.Param;
 import feign.RequestLine;
 
-import java.util.List;
 
 public interface LocalHeroChApi {
-    @RequestLine("GET /index.php?rest_route={restRoute}")
-    List<LocalHeroPost> findAll(@Param("restRoute") String restRoute);
+
+	@RequestLine("GET /wp-json/wp/v2/posts?page={page}&per_page={size}")
```
Comment (pavax) [inline integrations/localhero/src/main/java/ch/stayhome/integrations/localhero/infrastructure/feign/LocalHeroChApi.java:12]: @lugipfupf I was not able to get it working with the rest_route approach AND the pagination

File: integrations/localhero/pom.xml (hunk 8 above)
Comment (pavax) [reply integrations/localhero/pom.xml]: I never used the complete in memory approach but used a in-memory h2 instead -> but you are right I'll revert these db related changes

File: integrations/localhero/src/main/java/ch/stayhome/integrations/localhero/infrastructure/feign/LocalHeroChApi.java (hunk 7 above)
Comment (lugipfupf) [reply integrations/localhero/src/main/java/ch/stayhome/integrations/localhero/infrastructure/feign/LocalHeroChApi.java]: I was unable get any posts this way from when they redesigned their site to point to bern.localhero.ch and winterthur.local-hero.ch, respectively. All I got back was an empty json. So I used the discovery HEAD requests on bern.local-hero.ch which responded with this link. I know they also did some work on their side these days, so this version does seem to work now.

**Review by lugipfupf — APPROVED** [review]

**Comment by lugipfupf** [comment]
I'll merge later tomorrow if you don't find the time

**Pull request merged**